python_modules/dagster/dagster/_core/execution/calculate_data_time.py

```diff
@@ -0,0 +1,125 @@
+from datetime import datetime, timezone
+from typing import AbstractSet, Mapping, Optional, Tuple
+
+from ..definitions.events import AssetKey
+from ..instance import DagsterInstance
+
+ROOT_DATA_TAG = ".dagster/root_data_ids"
+
+
+def get_upstream_materialization_times_for_key(
+    instance: DagsterInstance,
+    asset_key: AssetKey,
+    upstream_asset_key_mapping: Mapping[str, AbstractSet[str]],
+) -> Mapping[str, Tuple[Optional[int], Optional[datetime]]]:
```

Comment:
Skimming this function definition and docstring, it's not immediately clear to me what entities the `str`, `Optional[int]`, and `Optional[datetime]` in the return value correspond to.